File: frontend/src/api/shopping.ts
import { authFetch } from './auth';

const API = '/api/v1';

export interface ShoppingItem {
  id: string;
  title: string;
  qty: string | null;
  bought: boolean;
  created_at: string;
  bought_at: string | null;
}

export async function listShopping(includeBought = true): Promise<ShoppingItem[]> {
  const r = await authFetch(`${API}/shopping?include_bought=${includeBought}`);
  if (!r.ok) throw new Error(`list failed: ${r.status}`);
  return r.json();
}

export async function createShopping(title: string, qty?: string | null): Promise<ShoppingItem> {
  const body: Record<string, unknown> = { title };
  if (qty) body.qty = qty;
  const r = await authFetch(`${API}/shopping`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error(`create failed: ${r.status}`);
  return r.json();
}

export async function updateShopping(
  id: string,
  changes: { title?: string; qty?: string | null; bought?: boolean },
): Promise<ShoppingItem> {
  const r = await authFetch(`${API}/shopping/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(changes),
  });
  if (!r.ok) throw new Error(`update failed: ${r.status}`);
  return r.json();
}

export async function deleteShopping(id: string): Promise<void> {
  const r = await authFetch(`${API}/shopping/${id}`, { method: 'DELETE' });
  if (!r.ok) throw new Error(`delete failed: ${r.status}`);
}

export async function clearBought(): Promise<{ removed: number }> {
  const r = await authFetch(`${API}/shopping/clear-bought`, { method: 'POST' });
  if (!r.ok) throw new Error(`clear failed: ${r.status}`);
  return r.json();
}

File: backend/cara/api/v1/shopping.py
"""Shopping list endpoints — auth, user-scoped."""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from cara.api.deps import get_current_user
from cara.models.user import User
from cara.schemas.shopping import (
    ShoppingItemCreate,
    ShoppingItemOut,
    ShoppingItemUpdate,
)
from cara.services import shopping as svc
from cara.services.family_bus import publish as fb_publish
from cara.store import get_session

router = APIRouter(prefix="/shopping", tags=["shopping"])


@router.get("", response_model=list[ShoppingItemOut])
async def list_shopping(
    include_bought: bool = True,
    user: User = Depends(get_current_user),  # noqa: B008
    session: AsyncSession = Depends(get_session),  # noqa: B008
) -> list[ShoppingItemOut]:
    rows = await svc.list_items(session, user_id=user.id, include_bought=include_bought)
    return [ShoppingItemOut.model_validate(r) for r in rows]


@router.post("", response_model=ShoppingItemOut, status_code=status.HTTP_201_CREATED)
async def create_shopping(
    body: ShoppingItemCreate,
    user: User = Depends(get_current_user),  # noqa: B008
    session: AsyncSession = Depends(get_session),  # noqa: B008
) -> ShoppingItemOut:
    item = await svc.create_item(session, user_id=user.id, title=body.title, qty=body.qty)
    out = ShoppingItemOut.model_validate(item)
    await fb_publish("shopping.created", user_id=user.id, payload=out.model_dump(mode="json"))
    return out


@router.patch("/{item_id}", response_model=ShoppingItemOut)
async def update_shopping(
    item_id: uuid.UUID,
    body: ShoppingItemUpdate,
    user: User = Depends(get_current_user),  # noqa: B008
    session: AsyncSession = Depends(get_session),  # noqa: B008
) -> ShoppingItemOut:
    fields = body.model_fields_set
    kwargs: dict = {}
    if "title" in fields:
        kwargs["title"] = body.title
    if "qty" in fields:
        kwargs["qty"] = body.qty if body.qty is not None else ""
    if "bought" in fields:
        kwargs["bought"] = body.bought
    item = await svc.update_item(session, item_id, user_id=user.id, **kwargs)
    if item is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "item not found")
    out = ShoppingItemOut.model_validate(item)
    await fb_publish("shopping.updated", user_id=user.id, payload=out.model_dump(mode="json"))
    return out


@router.delete("/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_shopping(
    item_id: uuid.UUID,
    user: User = Depends(get_current_user),  # noqa: B008
    session: AsyncSession = Depends(get_session),  # noqa: B008
) -> None:
    ok = await svc.delete_item(session, item_id, user_id=user.id)
    if not ok:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "item not found")
    await fb_publish("shopping.deleted", user_id=user.id, payload={"id": str(item_id)})


@router.post("/clear-bought")
async def clear_bought(
    user: User = Depends(get_current_user),  # noqa: B008
    session: AsyncSession = Depends(get_session),  # noqa: B008
) -> dict:
    n = await svc.clear_bought(session, user_id=user.id)
    return {"removed": n}
